### PR TITLE
[SPARK-37695][CORE][SHUFFLE] Skip diagnosis ob merged blocks from push-based shuffle

### PR DESCRIPTION
**Pull request by pan3793**
### What changes were proposed in this pull request?

Skip diagnosis ob merged blocks from push-based shuffle

### Why are the changes needed?

Because SPARK-36284 has not been addressed yet, skip it to suppress exceptions.

```
21/12/19 18:46:37 WARN TaskSetManager: Lost task 166.0 in stage 1921.0 (TID 138855) (beta-spark5 executor 218): java.lang.AssertionError: assertion failed: Expected ShuffleBlockId, but got shuffleChunk_464_0_5645_0
	at scala.Predef$.assert(Predef.scala:223)
	at org.apache.spark.storage.ShuffleBlockFetcherIterator.diagnoseCorruption(ShuffleBlockFetcherIterator.scala:1043)
	at org.apache.spark.storage.BufferReleasingInputStream.$anonfun$tryOrFetchFailedException$1(ShuffleBlockFetcherIterator.scala:1308)
	at scala.Option.map(Option.scala:230)
	at org.apache.spark.storage.BufferReleasingInputStream.tryOrFetchFailedException(ShuffleBlockFetcherIterator.scala:1307)
	at org.apache.spark.storage.BufferReleasingInputStream.read(ShuffleBlockFetcherIterator.scala:1293)
	at java.io.BufferedInputStream.fill(BufferedInputStream.java:246)
	at java.io.BufferedInputStream.read(BufferedInputStream.java:265)
	at java.io.DataInputStream.readInt(DataInputStream.java:387)
	at org.apache.spark.sql.execution.UnsafeRowSerializerInstance$$anon$2$$anon$3.readSize(UnsafeRowSerializer.scala:113)
	at org.apache.spark.sql.execution.UnsafeRowSerializerInstance$$anon$2$$anon$3.<init>(UnsafeRowSerializer.scala:120)
	at org.apache.spark.sql.execution.UnsafeRowSerializerInstance$$anon$2.asKeyValueIterator(UnsafeRowSerializer.scala:110)
	at org.apache.spark.shuffle.BlockStoreShuffleReader.$anonfun$read$2(BlockStoreShuffleReader.scala:98)
	at scala.collection.Iterator$$anon$11.nextCur(Iterator.scala:486)
	at scala.collection.Iterator$$anon$11.hasNext(Iterator.scala:492)
	at scala.collection.Iterator$$anon$10.hasNext(Iterator.scala:460)
	at org.apache.spark.util.CompletionIterator.hasNext(CompletionIterator.scala:31)
	at org.apache.spark.InterruptibleIterator.hasNext(InterruptibleIterator.scala:37)
	at scala.collection.Iterator$$anon$10.hasNext(Iterator.scala:460)
	at org.apache.spark.sql.catalyst.expressions.GeneratedClass$GeneratedIteratorForCodegenStage9.sort_addToSorter_0$(Unknown Source)
	at org.apache.spark.sql.catalyst.expressions.GeneratedClass$GeneratedIteratorForCodegenStage9.processNext(Unknown Source)
	at org.apache.spark.sql.execution.BufferedRowIterator.hasNext(BufferedRowIterator.java:43)
	at org.apache.spark.sql.execution.WholeStageCodegenExec$$anon$1.hasNext(WholeStageCodegenExec.scala:759)
	at org.apache.spark.sql.catalyst.expressions.GeneratedClass$GeneratedIteratorForCodegenStage10.smj_findNextJoinRows_0$(Unknown Source)
	at org.apache.spark.sql.catalyst.expressions.GeneratedClass$GeneratedIteratorForCodegenStage10.processNext(Unknown Source)
	at org.apache.spark.sql.execution.BufferedRowIterator.hasNext(BufferedRowIterator.java:43)
	at org.apache.spark.sql.execution.WholeStageCodegenExec$$anon$2.hasNext(WholeStageCodegenExec.scala:778)
	at scala.collection.Iterator$$anon$10.hasNext(Iterator.scala:460)
	at org.apache.spark.shuffle.sort.UnsafeShuffleWriter.write(UnsafeShuffleWriter.java:179)
	at org.apache.spark.shuffle.ShuffleWriteProcessor.write(ShuffleWriteProcessor.scala:59)
	at org.apache.spark.scheduler.ShuffleMapTask.runTask(ShuffleMapTask.scala:99)
	at org.apache.spark.scheduler.ShuffleMapTask.runTask(ShuffleMapTask.scala:52)
	at org.apache.spark.scheduler.Task.run(Task.scala:136)
	at org.apache.spark.executor.Executor$TaskRunner.$anonfun$run$3(Executor.scala:507)
	at org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:1468)
	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:510)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748) 
```

### Does this PR introduce _any_ user-facing change?

Yes, suppress the exceptions.

### How was this patch tested?

Run 1T TPCDS manually.

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/storage/ShuffleBlockFetcherIterator.scala [hunk 1]
```diff
@@ -1301,9 +1301,12 @@ private class BufferReleasingInputStream(
...
       case e: IOException if detectCorruption =>
-        val diagnosisResponse = checkedInOpt.map { checkedIn =>
-          iterator.diagnoseCorruption(checkedIn, address, blockId)
-        }
+        val diagnosisResponse = checkedInOpt
+          // TODO SPARK-36284 Add shuffle checksum support for push-based shuffle
+          .filterNot { _ => blockId.isShuffleChunk }
+          .map { checkedIn =>
+            iterator.diagnoseCorruption(checkedIn, address, blockId)
+          }
```
Comment (Ngone51) [inline core/src/main/scala/org/apache/spark/storage/ShuffleBlockFetcherIterator.scala:1309]: Shall we do the check inside `diagnoseCorruption()` to ensure other calls on it are also safe?

Comment (pan3793) [reply core/src/main/scala/org/apache/spark/storage/ShuffleBlockFetcherIterator.scala]: Updated

File: core/src/main/scala/org/apache/spark/storage/ShuffleBlockFetcherIterator.scala [hunk 2]
```diff
@@ -1037,39 +1037,48 @@ final class ShuffleBlockFetcherIterator(
...
-        s"Block $blockId is corrupted due to $otherCause"
+          case otherCause =>
+            s"Block $blockId is corrupted due to $otherCause"
+        }
+        logInfo(s"Finished corruption diagnosis in $duration ms. $diagnosisResponse")
+        diagnosisResponse
+      case shuffleBlockChunk: ShuffleBlockChunkId =>
+        // TODO SPARK-36284 Add shuffle checksum support for push-based shuffle
+        val diagnosisResponse = s"BlockChunk $shuffleBlockChunk is corrupted but corruption " +
+          s"diagnosis failed due to lack of shuffle checksum support for push-based shuffle."
```
Comment (Ngone51) [inline core/src/main/scala/org/apache/spark/storage/ShuffleBlockFetcherIterator.scala:1076]: "...diagnosis is skipped due to..."?

File: core/src/main/scala/org/apache/spark/storage/ShuffleBlockFetcherIterator.scala [hunk 3]
```diff
@@ -1037,39 +1037,48 @@ final class ShuffleBlockFetcherIterator(
       blockId: BlockId): String = {
     logInfo("Start corruption diagnosis.")
     val startTimeNs = System.nanoTime()
```
Comment (Ngone51) [inline core/src/main/scala/org/apache/spark/storage/ShuffleBlockFetcherIterator.scala:1039]: This line could move to `ShuffleBlockId` case only.

**Comment by Ngone51** [comment]
LGTM excepts 2 minor comments.

**Comment by Ngone51** [comment]
cc @mridulm

File: core/src/main/scala/org/apache/spark/storage/ShuffleBlockFetcherIterator.scala (hunk 2 above)
Comment (pan3793) [reply core/src/main/scala/org/apache/spark/storage/ShuffleBlockFetcherIterator.scala]: Updated

File: core/src/main/scala/org/apache/spark/storage/ShuffleBlockFetcherIterator.scala (hunk 3 above)
Comment (pan3793) [reply core/src/main/scala/org/apache/spark/storage/ShuffleBlockFetcherIterator.scala]: Updated

**Comment by yaooqinn** [comment]
ok to test

**Review by yaooqinn — APPROVED** [review]
LGTM

**Review by Ngone51 — APPROVED** [review]

**Comment by Ngone51** [comment]
Thanks, merged to master, branch-3.2.

**Pull request closed without merging**